src/metabase/query_processor/middleware/cache.clj

```diff
@@ -103,13 +103,13 @@
        (add-object-to-cache! (if (map? result)
                                (m/dissoc-in result [:data :rows])
                                {}))
-       (let [duration-ms  (- (System/currentTimeMillis) start-time)
-             min-duration (:min-duration strategy 0)
-             eligible?    (and @has-rows?
-                               (> duration-ms min-duration))]
+       (let [duration-ms     (/ (- (System/nanoTime) start-time-ns) 1e6)
```

Comment:
If we call `math/ceil` here do we need to relax the schemas?

Reply:
Well, yep, because it could be under a millisecond anyway. So it's either relaxing schema to take 0 or having `>=` in the check - which is going to be effectively the same, but harder for people to realize. So I think relaxing schema makes more sense here.

Reply:
But 1e6 is a float so wouldn't this number always be positive? I think in some JVM implements `nanoTime` is strictly monotonic but not sure.

Reply:
You could also take the max with 1 😆 - not sure whether I'm joking.

Reply:
Well I hope it's always positive, since it's a duration of how long query is running. But for test queries, which are really simple, it could be faster than a single millisecond - which is why for tests I really want 0 as minimum duration just to check caching works properly.

Reply:
[System.nanoTime()'s Javadoc](https://docs.oracle.com/javase/7/docs/api/java/lang/System.html#nanoTime()) makes it clear that all observed invocations within a JVM (that is, across all threads) are monotonic.

Reply:
Yeah, but not *strictly* monotonic. I did a bit of searching now and I think it's quite likely to return the same value if timing really fast things, as the actual precision of the timer might only be in milliseconds.

Reply:
Ehhh good enough for the purpose 😁

Reply:
We're talking past each other a bit I think, but we'll get there 😆 

There's a part of me that thinks the implementation should be like this.

```clojure
(defn- since-ms [timer] (max 1 (math/ceil (/ (- (System/nanoTime) timer) 1e6))))
```

Since 99.999999999% of timing code doesn't do this though, probably a bad idea though. Sad.

Reply:
> We're talking past each other a bit I think, but we'll get there 😆

I'm seeing your comments _eventually_ and in different order here and in emails and it melts my brain 🤣